### PR TITLE
Documentation for multiple classes

### PR DESCRIPTION
**Pull request by sigurdurb**
Documentation on how to setup Multiple classes with No authentication, Jupyterhub authentication and Custom authentication.

Revisions welcome.
@jhamrick @BertR @damianavila

### CODE REVIEW

**Comment by jhamrick** [comment]
One thing I'm thinking it might be useful to add to this as well is a table specifying the configuration options for different extensions for different users when you have separate services for the formgraders. Something like this:

|                   | Students | Instructors | Formgraders |
|-------------------|----------|-------------|-------------|
| Create Assignment | no       | no          | yes         |
| Assignment List   | yes      | yes       | no          |
| Formgrader        | no       | no          | yes         |
| Course List       | no       | yes         | no          |

**Comment by sigurdurb** [comment]
Yes! Where to enable which extension. Will put it in there.

File: nbgrader/docs/source/configuration/jupyterhub_config.rst [hunk 1]
```diff
@@ -195,3 +200,135 @@ if students may be enrolled in multiple courses, you need to provide them a
...
+For example, you could have users ``grader-course101`` and
+``grader-course123``. Your JupyterHub config would then look something like
+this:
+
+.. code:: python
+
+    c = get_config()
+
+    # Our user list
+    c.Authenticator.whitelist = [
```
Comment (jhamrick) [inline nbgrader/docs/source/configuration/jupyterhub_config.rst]: I think `grader-course101` and `grader-course123` need to be added to the whitelist as well.

File: nbgrader/docs/source/configuration/jupyterhub_config.rst [hunk 2]
```diff
@@ -195,3 +200,135 @@ if students may be enrolled in multiple courses, you need to provide them a
...
+JupyterHub Authentication
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+
+With the advent of JupyterHubAuthPlugin students who don't have a specified course_id won't see all courses anymore, just the ones they have been added to. 
+
+Requirements for using ``JupyterHubAuthPlugin``: 
+
+* Jupyterhub > 0.8
+* Activating the JupyterHubAuthPlugin requires you to add it as an authentication plugin class into ``/etc/jupyter/nbgrader_config.py``
```
Comment (jhamrick) [inline nbgrader/docs/source/configuration/jupyterhub_config.rst]: Maybe say this instead:

```suggestion
* Activating the JupyterHubAuthPlugin requires you to add it as an authentication plugin class into the ``nbgrader_config.py`` for both the formgrader services and instructor accounts.
```

because other people may have their `nbgrader_config.py` files set up differently and paths are different on different systems. However I do want to create a separate piece of documentation (in a separate PR probably) about how to deal with different `nbgrader_config.py` files.

File: nbgrader/docs/source/configuration/jupyterhub_config.rst [hunk 3]
```diff
@@ -195,3 +200,135 @@ if students may be enrolled in multiple courses, you need to provide them a
...
+
+To make your own custom authentication such as through an LTI you could start by making a method that inherits the Authenticator class, which is a plugin for different authentication methods.
+
+There are now three authentication classes:
+
+* ``NoAuthPlugin`` : The default old behaviour. Using this plugin will allow any user to any course if he does not have a course_id in his nbgrader_config. This is still the default behaviour so no need to specify it in ``etc/jupyter/nbgrader_config.py``
+
+* ``JupyterHubAuthPlugin`` : Uses the Jupyterhub groups part of the Jupyterhub API for authentication.
+
+* ``Authenticator``: Inherit this class to make your own Authenticator plugin, thought of as a way to enable LTI use cases.
```
Comment (jhamrick) [inline nbgrader/docs/source/configuration/jupyterhub_config.rst]: Actually this isn't quite right---to have your own plugin you have to inherit from `BaseAuthPlugin`. The `Authenticator` class is the object which is configurable to have different plugins.

Could you also add here some documentation about the methods that have to be overwritten? These are `get_student_courses`, `add_student_to_course`, and `remove_student_from_course`. You can do this automatically with sphinx directives, something like this:

```
.. currentmodule:: nbgrader.auth

.. autoclass:: BaseAuthPlugin

    .. automethod:: get_student_courses

    .. automethod:: add_student_to_course

    .. automethod:: remove_student_from_course
```

File: nbgrader/docs/source/user_guide/third_party.rst [hunk 4]
```diff
@@ -50,6 +50,9 @@ Libraries, Tools, and Platforms
 Presentations and Educational Materials
 ---------------------------------------
 
+`Teaching and Learning with Jupyter Book <https://jupyter4edu.github.io/jupyter-edu-book/>`_
```
Comment (jhamrick) [inline nbgrader/docs/source/user_guide/third_party.rst]: Hmm, looks like this snuck in here because I think you merged master into your branch. I am going to merge master into the `multiple-classes-dev` branch and then you'll need to do another merge (or rebase) from `multiple-classes-dev` into your branch so these unrelated changes get dropped.

Comment (jhamrick) [reply nbgrader/docs/source/user_guide/third_party.rst]: Actually, I suspect that's going to cause a lot of merge conflicts, so I will actually leave it as is for now and do the merge later. It's ok if these changes sneak in for now :)

File: nbgrader/docs/source/configuration/jupyterhub_config.rst (hunk 3 above)
Comment (sigurdurb) [reply nbgrader/docs/source/configuration/jupyterhub_config.rst]: I added a normal rst table for the methods that need to be overridden, I could not get the extensions in nbgrader/docs/source/conf.py to work with this. There was no currentmodule extensions found, I tried adding autosummary, it accepted that extension but I had some errors.

File: nbgrader/docs/source/configuration/jupyterhub_config.rst (hunk 1 above)
Comment (sigurdurb) [reply nbgrader/docs/source/configuration/jupyterhub_config.rst]: Added to the whitelist

File: nbgrader/docs/source/configuration/jupyterhub_config.rst [hunk 5]
```diff
@@ -195,3 +199,155 @@ if students may be enrolled in multiple courses, you need to provide them a
...
+* ``BaseAuthPlugin``: Inherit this class when implementing your own plugin, thought of as a way to enable LTI use cases. This class is never called directly.
+* ``NoAuthPlugin``: The default old behaviour. Using this plugin will allow any user to any course if he does not have a course_id in his nbgrader_config. This is still the default behaviour so no need to specify it in ``etc/jupyter/nbgrader_config.py``
+
+* ``JupyterHubAuthPlugin``: Uses the Jupyterhub groups part of the Jupyterhub API for authentication.
+
+* ``Authenticator``: Configurable for different plugins.
+
+
+
+**BaseAuthPlugin methods to override for custom authentication**
```
[the remark this answers is not in the document]
Comment (sigurdurb) [reply nbgrader/docs/source/configuration/jupyterhub_config.rst]: Ahh yes, now it works.

**Comment by sigurdurb** [comment]
@jhamrick Got the table in, we could make a separate chapter just for the extensions but this should be good for now.

**Comment by jhamrick** [comment]
Hooray, the tests pass! Merging now 🙂 Thanks for all your work on this @sigurdurb!
